client/src/graphics/AnimationLoader.lua

```diff
@@ -0,0 +1,441 @@
+
+  -- First all files are loaded from JSON and indexed
+  AnimationLoader.readFileAndCollectIdsRecursively(filePath, loadedFileTables, idLookupTable)
+
+  local rootData = loadedFileTables[filePath]
+
+  -- Then all references are resolved with overrides applied
+  for _, drawable in ipairs(rootData.drawables) do
+    AnimationLoader.resolveRefsRecursively(drawable, loadedFileTables, idLookupTable, {}, filePath, rootPath)
+  end
+
+  -- Finally once the full data structure is setup, it is converted into separate drawable objects.
+  local drawables = {}
+  if rootData then
+    for _, data in ipairs(rootData.drawables) do
+      local drawable = loadDrawables(rootPath, data)
+      if drawable then
+        drawables[#drawables+1] = drawable
+      end
+    end
+  end
+  return drawables
+end
+
+function AnimationLoader.stopFluxTweensOnDrawable(drawable)
+    for _, tween in ipairs(drawable.fluxTweens) do
+      tween:stop()
+    end
+    for _, child in ipairs(drawable.children) do
+      AnimationLoader.stopFluxTweensOnDrawable(child)
+    end
+end
+
+-- DRAWING CODE
+
+-- Shader used for clipping using a stencil pass
+local alphaDiscardShader = love.graphics.newShader([[
+    vec4 effect(vec4 tintColor, Image tex, vec2 texCoord, vec2 screenCoord)
+    {
+        vec4 pixelColor = Texel(tex, texCoord);
+
+        if (pixelColor.a * tintColor.a <= 0.0)
+        {
+            discard;
+        }
+
+        return vec4(0.0);
+    }
+]])
+
+function AnimationLoader.objectTransform(obj)
+  local ax, ay = AnimationLoader.anchorOffset(obj, obj.anchor)
+  return obj.x - ax, obj.y - ay, obj.rotation, obj.xScale, obj.yScale
+end
+
+
+function AnimationLoader.drawNode(d, parent)
+    love.graphics.push("all")
+
+    local px, py = AnimationLoader.anchorOffset(d, d.pivot)
+    local wx, wy, wrot, xScale, yScale = AnimationLoader.objectTransform(d)
+
+    love.graphics.translate(wx, wy)
+
+    love.graphics.translate(px, py)
+    love.graphics.rotate(wrot)
+    love.graphics.scale(xScale, yScale)
+    love.graphics.translate(-px, -py)
+
+    if d.texture then
+      if d.stencil then
+        assert(parent, "To use a stencil you need siblings")
+        love.graphics.stencil(function()
+          love.graphics.setShader(alphaDiscardShader)
+          for _, sibling in ipairs(parent.children) do
+            if sibling == d then
+              break
+            end
+            AnimationLoader.drawNode(sibling, nil)
+          end
+          love.graphics.setShader()
+        end, "replace", 1, false)
+        love.graphics.setStencilTest("equal", 1)
+      end
+
+      love.graphics.setBlendMode(d.blendMode, d.alphaMode)
+      love.graphics.setColor(d.tint[1], d.tint[2], d.tint[3], d.alpha)
+      if d.quad then
+        d.quad:setViewport(d.scrollX, d.scrollY, d.width, d.height)
+        love.graphics.draw(d.texture, d.quad, 0, 0)
+      else
+        love.graphics.draw(d.texture, 0, 0)
+      end
+      love.graphics.setStencilTest()
```

Comment:
`setStencilTest` and `setBlendMode` both break love's autobatching so they should only be called if it's actually necessary.
Given the setup only the stars significantly benefit but just checking against `d.stencil` and the output of `getBlendMode` is already 20% less draw time for the main menu.